specs/interop/token-bridging.md

```diff
@@ -2,96 +2,296 @@
+function sendERC20(uint256 _amount, uint256 _chainId) external returns (bool) {
+  return sendERC20To(msg.sender, _amount, _chainId);
+}
 
-  function bridgeERC20(uint256 _amount, uint256 _chainId) external {
-    bridgeERC20To(msg.sender, _amount, _chainId);
-  }
+function sendERC20To(address _to, uint256 _amount, uint256 _chainId) public returns (bool success) {
+  _burn(msg.sender, _amount);
+  bytes memory _message = abi.encodeWithSignature("finalizeSendERC20(address,uint256)", _to, _amount);
+  _sendMessage(_chainId, _message);
+  return true
+}
 
-  function bridgeERC20To(address _to, uint256 _amount, uint256 _chainId) public {
-    _burn(msg.sender, amount);
+function sendERC20To(address _to, uint256 _amount, uint256 _chainId, bytes memory _data) public returns (bool success) {
+  _burn(msg.sender, _amount);
+  bytes memory _message = abi.encodeWithSignature("finalizeSendERC20(address,uint256,bytes)", _to, _amount, _data);
+  _sendMessage(_chainId, _message);
+  return true
+}
 
-    L2ToL2CrossDomainMessenger.sendMessage({
-      _destination: chainId,
-      _target: address(this),
-      _message: abi.encodeCall(this.finalizeBridgeERC20, (to, amount))
-    });
-  }
+function finalizeSendERC20(address _to, uint256 _amount) external {
+  // this will be a modifier, inline here for clarity
+  require(msg.sender == address(L2ToL2CrossChainMessenger));
+  require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+  //
+  _mint(_to, _amount);
+}
 
-  function finalizeBridgeERC20(address _to, uint256 _amount) external {
-    require(msg.sender == address(L2ToL2CrossChainMessenger));
-    require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+function finalizeSendERC20(address _to, uint256 _amount, bytes memory _data) external {
+  require(msg.sender == address(L2ToL2CrossChainMessenger));
+  require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+  _mint(_to, _amount);
+  (bool success, ) = _to.call(_data);
+  require(success, "External call failed");
+}
 
-    _mint(to, amount);
-  }
+/////// Internal functions ///////////
+function _sendMessage(uint256 _destination, bytes memory _data) internal virtual {
+  L2ToL2CrossDomainMessenger.sendMessage(_destination, address(this), _data);
+}
+```
+
+Note: Some other naming options that were considered were `bridgeERC20`, `send()`, `xTransfer()`, `Transfer` (with a chainId parameter).
+
+## Invariants
+
+Besides the ERC20 invariants, the SuperchainERC20 will require the following interop specific properties:
+
+- Conservation of finalized `totalSupply`: The minted `amount` in `finalizeSendERC20()` should match the `amount` that was burnt in `sendERC20To()`, as long as target chain has the initiating chain in the dependency set. This implies that finalized cross-chain transactions will conserve the sum of `totalSupply` for each chain in the Superchain.
```

Comment:
> as long as target chain has the initiating chain in the dependency set

What happens if the initiating chain is not in the dependency set? Do the send methods need to validate that the chain ID is in the dependency set, or does the executing message just never get included? Who is expected to call the finalize method?

Reply:
Thats a very good open question, and its something we have to discuss further. We considered adding rollback functionalities for these cases, but we think its out of scope for the moment.
For that reason, I included the dependancy set clarification into the invariant. If there were a way to handle that scenario, the invariant could be extended.